### PR TITLE
Fixed the problem with products create/update via REST API requests when request body data is sent as "xml"

### LINKED ISSUE #28155: [Issue] Fixed the problem with products create/update via REST API requests when request body data is sent as "xml"
This issue is automatically created based on existing pull request: magento/magento2#28140: Fixed the problem with products create/update via REST API requests when request body data is sent as "xml"

---------
Fixed the problem with products create/update via REST API requests when request body data is sent as "xml"

### Description (*)
After updating Magento to v. 2.3.5 there is no more possibility to send REST API requests for creating/updating products with "xml" request body. Every request leads to a response with an error: "Decoding error: Unable to unserialize value. Error:..."
The problem appeared after adding the class "Magento\Catalog\Model\Product\Webapi\ProductOutputProcessor" which additionally processes product data during preparing API response.
This class uses deserializer class ("Magento\Framework\Webapi\Rest\Request\DeserializerInterface"  inherited class instance) for the processing (converting) request body (interface injected in the constructor).
Different deserializer class configured (via di.xml) for REST and SOAP scopes: "Magento\Framework\Webapi\Rest\Request\Deserializer\Json" (for REST) and  Magento\Framework\Webapi\Rest\Request\Deserializer\XML (for SOAP).
https://github.com/magento/magento2/commit/f58ddb5b2996d4069a0c79bf693d7f92bbb2e5e5#diff-8f3c9c53a4d283c4a677a3bfd544a363R40
Such configuration leads to parsing errors in product REST API requests when product request data sent as XML because   ProductOutputProcessor will always use Json deserializer for converting request body data without checking origin request content-type (which could be XML). 
For fixing this I have implemented custom deserializer class for products rest requests:
*Magento\Catalog\Model\Product\Webapi\Rest\RequestTypeBasedDeserializer*
This class first checks content-type for request body and then, using deserializer factory, get needed deserializer instance and use it for converting request body data.

-->

### Manual testing scenarios (*)
**Test on Magento version 2.3.5 or using 2.4-develop branch**
1. Sent product create/update request with XML request body and "Content-Type: application/xml" header. 
**Example:** 
```
curl --location --request POST 'https://magento-site/rest/all/V1/products' \
--header 'Authorization: Bearer someencodedtoken' \
--header 'Content-Type: application/xml' \
--data-raw '<?xml version='\''1.0'\'' encoding='\''UTF-8'\''?>
<products>
    <product>
        <sku>TRAKTOR_KONTROL_BAG_1</sku>
        <name>TRAKTOR_KONTROL_BAG_1</name>
        <weight>22</weight>
        <attribute_set_id>4</attribute_set_id>
        <status>2</status>
      </product>
</products>
```
2. Got an error in response with the text: "Decoding error: Unable to unserialize value. Error:..." 

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### PR DESCRIPTION
**Pull request by Neos2007**
Fixed the problem with products create/update via REST API requests when request body data is sent as "xml"

### Description (*)
After updating Magento to v. 2.3.5 there is no more possibility to send REST API requests for creating/updating products with "xml" request body. Every request leads to a response with an error: "Decoding error: Unable to unserialize value. Error:..."
The problem appeared after adding the class "Magento\Catalog\Model\Product\Webapi\ProductOutputProcessor" which additionally processes product data during preparing API response.
This class uses deserializer class ("Magento\Framework\Webapi\Rest\Request\DeserializerInterface"  inherited class instance) for the processing (converting) request body (interface injected in the constructor).
Different deserializer class configured (via di.xml) for REST and SOAP scopes: "Magento\Framework\Webapi\Rest\Request\Deserializer\Json" (for REST) and  Magento\Framework\Webapi\Rest\Request\Deserializer\XML (for SOAP).
https://github.com/magento/magento2/commit/f58ddb5b2996d4069a0c79bf693d7f92bbb2e5e5#diff-8f3c9c53a4d283c4a677a3bfd544a363R40
Such configuration leads to parsing errors in product REST API requests when product request data sent as XML because   ProductOutputProcessor will always use Json deserializer for converting request body data without checking origin request content-type (which could be XML). 
For fixing this I have implemented custom deserializer class for products rest requests:
*Magento\Catalog\Model\Product\Webapi\Rest\RequestTypeBasedDeserializer*
This class first checks content-type for request body and then, using deserializer factory, get needed deserializer instance and use it for converting request body data.

-->

### Manual testing scenarios (*)
**Test on Magento version 2.3.5 or using 2.4-develop branch**
1. Sent product create/update request with XML request body and "Content-Type: application/xml" header. 
**Example:** 
```
curl --location --request POST 'https://magento-site/rest/all/V1/products' \
--header 'Authorization: Bearer someencodedtoken' \
--header 'Content-Type: application/xml' \
--data-raw '<?xml version='\''1.0'\'' encoding='\''UTF-8'\''?>
<products>
    <product>
        <sku>TRAKTOR_KONTROL_BAG_1</sku>
        <name>TRAKTOR_KONTROL_BAG_1</name>
        <weight>22</weight>
        <attribute_set_id>4</attribute_set_id>
        <status>2</status>
      </product>
</products>
```
2. Got an error in response with the text: "Decoding error: Unable to unserialize value. Error:..." 

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### Resolved issues:
1. [x] resolves magento/magento2#28155: Fixed the problem with products create/update via REST API requests when request body data is sent as "xml"

### CODE REVIEW

**Comment by ihor-sviziev** [comment]
@magento create issue

**Comment by ihor-sviziev** [comment]
Hi @Neos2007,
Could you report the issue with all exact steps to reproduce? Does this issue appears only on catalog, or somewhere else?

**Comment by Neos2007** [comment]
@ihor-sviziev the problem related and reproducible only for the product create/update requests because of logic "Magento\Catalog\Model\Product\Webapi\ProductOutputProcessor" (as I mentioned in the ticket description) class which was added in Magento v. 2.3.5.

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @Neos2007,
I just reviewed your changes - and commit https://github.com/magento/magento2/commit/f58ddb5b2996d4069a0c79bf693d7f92bbb2e5e5#diff-d0dca5b5a6dc36b5f088d8d9c9630db6R56 where this class was introduced.

Could you explain, why do you think it's good idea to create new class RequestTypeBasedDeserializer that will be available only in Catalog rather than updating `ProductOutputProcessor` class to use `DeserializerFactory` to retrieve correct object and then execute `deserialize` method on it? Seems like issue just inside only one class, I don't think we actually have any benefit from introducing new class.

Could you also sign Adobe CLA, otherwise we can't accept your PR 😢

**Comment by Neos2007** [comment]
@ihor-sviziev I moved this logic in separate deserializer class because determining needed deserializer (XML, JSON) must be done only for rest requests. But "ProductOutputProcessor" class used also in SOAP requests and it requires to be injected XML parser.

And also I have some problems with sign Adobe CLA. I Signed it. But status was not changed (I tried to close and reopen PR after I signed).

**Comment by ihor-sviziev** [comment]
@Neos2007 could you write me in Slack? It will be easier to figure out why it fails

**Review by ihor-sviziev — DISMISSED** [review]
Looks good! Approved ✔

Would be good to cover such case with API Functional test, but as we discussed with @Neos2007 - they're not supporting XML in REST api 😢
https://github.com/magento/magento2/blob/d551d1e3f70f8ae09c4de6958125f22b855d5569/dev/tests/api-functional/framework/Magento/TestFramework/TestCase/Webapi/Adapter/Rest.php#L79

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-7530](https://jira.corp.magento.com/browse/ENGCOM-7530) has been created to process this Pull Request
:eight_spoked_asterisk: @ihor-sviziev, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Review by nuzil — COMMENTED** [review]
I love it! 
Thanks @ihor-sviziev for fast processing. Quite critical issue currently.

**Comment by ihor-sviziev** [comment]
@magento create issue

**Comment by ihor-sviziev** [comment]
I've reported an issue to provide an ability to run XML in API Functional tests
https://github.com/magento/magento2/issues/28156

**Comment by slavvka** [comment]
@Neos2007 @ihor-sviziev @nuzil PO decided to deliver it to `2.4.0` if it is completed (with testing) within a week. Please note that I changed a base to `2.4.0-develop` so it should be retested. Also keep in mind that today or tomorrow there will be merged support for PHP 7.4 and lots of files will be changed so it need to be compatible with it.

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

**Manual testing scenario:**

1. Sent product create request with XML request body and "Content-Type: application/xml" header.
Screenshot from 2020-05-12 13-35-29

**Before:** 
:heavy_multiplication_x:  Got an error: "**Decoding error: Unable to unserialize value. Error:...**"
Screenshot from 2020-05-08 16-10-15

**After:**
:heavy_check_mark: 200 Response. The product was successfully created. 
Screenshot from 2020-05-12 13-35-47

2. Sent product update request with XML request body (add `<price>321</price>` for ex.)

**After:** :heavy_check_mark: The product was successfully updated. 

Screenshot from 2020-05-12 13-37-34

3. Create another product with JSON request body and "Content-Type: application/json" header.
Screenshot from 2020-05-12 13-47-07

**After:** The product was successfully created.

**Review by slavvka — CHANGES_REQUESTED** [review]
@Neos2007 Please merge the latest mainline and re-run tests

**Comment by Neos2007** [comment]
@magento run all tests

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @Neos2007,
Please fix compatibility issues with phpunit 9

File: app/code/Magento/Catalog/Test/Unit/Model/Product/Webapi/Rest/RequestTypeBasedDeserializerTest.php [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+     * @var DeserializerFactory|MockObject
+     */
+    private $deserializeFactoryMock;
+
+    /**
+     * @var Request|MockObject
+     */
+    private $requestMock;
+
+    public function setUp()
```
Comment (ihor-sviziev) [inline app/code/Magento/Catalog/Test/Unit/Model/Product/Webapi/Rest/RequestTypeBasedDeserializerTest.php]:
```suggestion
    public function setUp(): void
```

File: app/code/Magento/Catalog/Test/Unit/Model/Product/Webapi/Rest/RequestTypeBasedDeserializerTest.php [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+        $serializer =  new SerializerJson();
+        return new DeserializerJson($decoder, $appStateMock, $serializer);
+    }
+
+    /**
+     * Creates XML Deserializer instance with some mocked parameters
+     *
+     * @return DeserializerXml
+     */
+    private function prepareXmlDeserializer()
```
Comment (ihor-sviziev) [inline app/code/Magento/Catalog/Test/Unit/Model/Product/Webapi/Rest/RequestTypeBasedDeserializerTest.php]:
```suggestion
    private function prepareXmlDeserializer(): DeserializerXml
```

File: app/code/Magento/Catalog/Test/Unit/Model/Product/Webapi/Rest/RequestTypeBasedDeserializerTest.php [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+            ]
+        ];
+    }
+
+    /**
+     * Creates Json Deserializer instance with some mocked parameters
+     *
+     * @return DeserializerJson
+     */
+    private function prepareJsonDeserializer()
```
Comment (ihor-sviziev) [inline app/code/Magento/Catalog/Test/Unit/Model/Product/Webapi/Rest/RequestTypeBasedDeserializerTest.php]:
```suggestion
    private function prepareJsonDeserializer(): DeserializerJson
```

File: app/code/Magento/Catalog/Test/Unit/Model/Product/Webapi/Rest/RequestTypeBasedDeserializerTest.php [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+        $this->requestMock->method('getContentType')
+            ->willReturn($contentType);
+        $this->deserializeFactoryMock->expects($this->any())
+            ->method('get')
+            ->with($contentType)
+            ->willReturn($deserializer);
+        $this->assertEquals($expectedResult, $this->requestTypeBasedDeserializer->deserialize($body));
+    }
+
+    public function getDeserializerDataProvider()
```
Comment (ihor-sviziev) [inline app/code/Magento/Catalog/Test/Unit/Model/Product/Webapi/Rest/RequestTypeBasedDeserializerTest.php]:
```suggestion
    public function getDeserializerDataProvider(): array
```

File: app/code/Magento/Catalog/Test/Unit/Model/Product/Webapi/Rest/RequestTypeBasedDeserializerTest.php [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+     * @param array $expectedResult
+     * @throws \Magento\Framework\Exception\InputException
+     * @throws \Magento\Framework\Webapi\Exception
+     */
+    public function testDeserialize(
+        $body,
+        $contentType,
+        DeserializerInterface $deserializer,
+        array $expectedResult
+    ) {
```
Comment (ihor-sviziev) [inline app/code/Magento/Catalog/Test/Unit/Model/Product/Webapi/Rest/RequestTypeBasedDeserializerTest.php]:
```suggestion
    public function testDeserialize(
        string $body,
        string $contentType,
        DeserializerInterface $deserializer,
        array $expectedResult
    ): void {
```

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

**Comment by ihor-sviziev** [comment]
Hi @Neos2007,
Could you update your PR with changes I've listed above or provide write access for maintainers, so I'll be able to update your branch myself?

**Comment by nuzil** [comment]
Hi @ihor-sviziev, done

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-7530](https://jira.corp.magento.com/browse/ENGCOM-7530) has been created to process this Pull Request

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed
The results the same as in [comment](https://github.com/magento/magento2/pull/28140#issuecomment-627265785)

**Review by sdzhepa — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @sdzhepa, thank you for the review.
[ENGCOM-7530](https://jira.corp.magento.com/browse/ENGCOM-7530) has been created to process this Pull Request

**Pull request merged**